### PR TITLE
Delete role and policy on apps3bucket deletion

### PR DESCRIPTION
**Pull request by jmoz**
## What

When you delete the apps3bucket relation it should also detach policy and role.

(Note: apologies on line length reformatting - I updated to 80 char after team decided on that length)

## How to review

1. Run tests.
2. Call it live and check.

### CODE REVIEW

File: control_panel_api/views.py [hunk 1]
```diff
@@ -58,6 +58,10 @@ class AppS3BucketViewSet(viewsets.ModelViewSet):
     queryset = AppS3Bucket.objects.all()
     serializer_class = AppS3BucketSerializer
 
+    def perform_destroy(self, instance):
```
Comment (andyhd) [inline control_panel_api/views.py]: why do all these method names start with `perform_`? isn't that redundant?

File: control_panel_api/services.py [hunk 2]
```diff
@@ -163,3 +166,16 @@ def bucket_delete(name):
     """Simple facade function for view to call on s3bucket delete
     Note we do not destroy the actual data, just the policies"""
     delete_bucket_policies(name)
+
+
+def apps3bucket_delete(apps3bucket):
+    """On delete we detach the corresponding s3 read/write policy from app role
```
Comment (andyhd) [inline control_panel_api/services.py]: This comment only describes _what_ the function does, it doesn't explain _why_

File: control_panel_api/services.py [hunk 3]
```diff
@@ -163,3 +166,16 @@ def bucket_delete(name):
     """Simple facade function for view to call on s3bucket delete
     Note we do not destroy the actual data, just the policies"""
     delete_bucket_policies(name)
+
+
+def apps3bucket_delete(apps3bucket):
```
Comment (andyhd) [inline control_panel_api/services.py]: Shouldn't this function be a method on `AppS3Bucket`?

File: control_panel_api/services.py [hunk 4]
```diff
@@ -138,8 +139,10 @@ def create_bucket(bucket_name):
 
 def create_bucket_policies(bucket_name):
     """Create readwrite and readonly policies for s3 bucket"""
-    aws.create_policy(_policy_name(bucket_name, readwrite=True), get_policy_document(bucket_name, True))
-    aws.create_policy(_policy_name(bucket_name, readwrite=False), get_policy_document(bucket_name, False))
+    aws.create_policy(_policy_name(bucket_name, readwrite=True),
+                      get_policy_document(bucket_name, True))
+    aws.create_policy(_policy_name(bucket_name, readwrite=False),
+                      get_policy_document(bucket_name, False))
```
Comment (andyhd) [inline control_panel_api/services.py]: Please avoid arbitrary indentation levels. Break the line after the open parenthesis and indent 4 spaces.
```python
aws.create_policy(
    _policy_name(...
```

File: control_panel_api/tests/test_views.py [hunk 5]
```diff
@@ -177,6 +177,15 @@ def test_delete(self):
...
         self.assertEqual(HTTP_404_NOT_FOUND, response.status_code)
 
+    @patch('control_panel_api.services.apps3bucket_delete')
+    def test_delete_calls_api(self, mock_apps3bucket_delete):
+        response = self.client.delete(
+            reverse('apps3bucket-detail', (self.apps3bucket_1.id,)))
+        self.assertEqual(HTTP_204_NO_CONTENT, response.status_code)
+
+        called_args, _ = mock_apps3bucket_delete.call_args
+        self.assertIsInstance(called_args[0], AppS3Bucket)
```
Comment (xoen) [inline control_panel_api/tests/test_views.py]: As discussed offline earlier I would personally be more specific here and check the function was called with the right instance, e.g. 

```python
called_args, _ = mock_apps3bucket_delete.call_args
self.assertEqual(called_args[0].app, self.apps3bucket_1.app)
self.assertEqual(called_args[0].s3bucket, self.apps3bucket_1.s3bucket)
```

...or something like that, you got the idea. But that's just me.

File: control_panel_api/views.py (hunk 1 above)
Comment (xoen) [reply control_panel_api/views.py]: This is not exactly apparent and I may be wrong but I think here we're overriding a method coming from Django Rest Framework (the beauty of inheritance 💃 )

File: control_panel_api/services.py (hunk 4 above)
Comment (jmoz) [reply control_panel_api/services.py]: This follows pep8, and is auto formatted, see first example

https://www.python.org/dev/peps/pep-0008/#indentation

Do we have a custom code style to follow I can reference?

File: control_panel_api/tests/test_views.py (hunk 5 above)
Comment (jmoz) [reply control_panel_api/tests/test_views.py]: Yep done.

File: control_panel_api/services.py (hunk 2 above)
Comment (jmoz) [reply control_panel_api/services.py]: Yep, also removed other ones.

File: control_panel_api/views.py (hunk 1 above)
Comment (jmoz) [reply control_panel_api/views.py]: @andyhd what @xoen said!

File: control_panel_api/services.py (hunk 4 above)
Comment (xoen) [reply control_panel_api/services.py]: This is what PEP8 says about indentation: https://www.python.org/dev/peps/pep-0008/#indentation

I tend to write stuff like this as:

```
aws.create_policy(
    _policy_name(bucket_name, readwrite=True),
    get_policy_document(bucket_name, True),
)
aws.create_policy(
    _policy_name(bucket_name, readwrite=False),
    get_policy_document(bucket_name, False),
)
```

which is valid from a PEP8-perspective and it's visually clear what's an argument of the function, etc...BUT, you two are the python ninjas.

File: control_panel_api/tests/test_views.py (hunk 5 above)
Comment (xoen) [reply control_panel_api/tests/test_views.py]: 🌟

File: control_panel_api/services.py [hunk 6]
```diff
@@ -15,35 +15,29 @@ def _policy_name(bucket_name, readwrite=False):
...
 
 def app_create(app_slug):
-    """Creates the IAM role for the given app"""
     _create_app_role(_app_role_name(app_slug))
 
 
 def _create_app_role(role_name):
-    """Creates the IAM role for the given app"""
-
-    # See: `sts:AssumeRole` required by kube2iam
```
Comment (xoen) [inline control_panel_api/services.py]: I think that comment was actually useful (it certainly saved me some research time in the lambda function version the other day!)

**Comment by xoen** [comment]
You may need to rebase (let me know if you need any input from me to resolve possible conflicts) but 👍

File: control_panel_api/services.py (hunk 6 above)
Comment (jmoz) [reply control_panel_api/services.py]: Added back

File: control_panel_api/services.py (hunk 4 above)
Comment (andyhd) [reply control_panel_api/services.py]: It is acceptable to PEP8, but I think PEP8 is wrong on this one.

Comment (jmoz) [reply control_panel_api/services.py]: Cool.  I would rather follow a strict set of rules than having to debate _personal opinion and preference_.

However I happen to agree on this preference which is also valid pep8 @xoen @andyhd so I'll try and hack my formatter to use hanging!

Comment (xoen) [reply control_panel_api/services.py]: What if PEP8 was wrong

**Pull request merged**